__tests__/systemCalls.spec.ts

```diff
@@ -413,4 +414,53 @@ describe('SystemCalls', () => {
 
     expect(obj2).toBeNull();
   });
+
+  it('should check authority using contract metadata', () => {
+    /* The expected behavior is that if the contract is called from an operation,
+     * signing the transaction should authorize.
+     * Otherwise authorization is granted if the caller is the requested contract
+     * or if the contract has an authorize overload and that overload returns true.
+     */
+
+    let mockAccountAuthorizationTrue = new MockVM.MockAuthority(authority.authorization_type.contract_call, mockAccount, true);
+    let mockAccountAuthorizationFalse = new MockVM.MockAuthority(authority.authorization_type.contract_call, mockAccount, false);
+
+    MockVM.setContractMetadata(new chain.contract_metadata_object(new Uint8Array(0), false, false, false, false));
+    MockVM.setCaller(new chain.caller_data(new Uint8Array(0), chain.privilege.user_mode));
+    MockVM.setEntryPoint(0xc3ab8ff1);
+    MockVM.setContractArguments(mockAccount);
+
+    // Case 1: Caller is an op and the transaction is signed. Should authorize.
+    MockVM.setAuthorities([mockAccountAuthorizationTrue]);
+    expect(System.checkAuthority(authority.authorization_type.contract_call, mockAccount)).toBe(true);
+
+    // Case 2: Caller is an op and the transaction is not signed. Should not authorize.
+    MockVM.setAuthorities([mockAccountAuthorizationFalse]);
+    expect(System.checkAuthority(authority.authorization_type.contract_call, mockAccount)).toBe(false);
+
+    // Case 3: Caller is another contract and there is no authorize override. Should not authorize.
+    MockVM.setCaller(new chain.caller_data(mockAccount2, chain.privilege.user_mode));
+    System.resetCache();
+
+    MockVM.setAuthorities([mockAccountAuthorizationFalse]);
```

Comment:
It should not authorize **even if the user signed the transaction**. Then it's better to set
```js
 MockVM.setAuthorities([mockAccountAuthorizationTrue]);
```
to show that the user signed the transaction and it still fails.

`MockVM.setAuthorities` is ambiguos because it is not possible to know if we are defining that the user signed the transaction or if the authorize function of user's contract authorized the operation. However, in this case we can assume it's the first case (user signed the transaction) because at the beginning you defined that the contract metadata is empty.